plasmapy/particles/particle_class.py

```diff
@@ -2482,3 +2484,181 @@ def molecule(
+
+    Notes
+    -----
+    If the quantity is not specified, then the corresponding values of Energy,
+    Frequency, Momentum and Wavelength will be set to ``numpy.nan`` in
+    the appropriate units.
+
+    Examples
+    --------
+    >>> from astropy import units as u
+    >>> from plasmapy.particles import Photon
+    >>> photon = Photon(5 * u.m)
+    >>> photon.wavelength
+    <Quantity 5. m>
+    >>> photon.energy
+    <Quantity 3.97289171e-26 J>
+    >>> photon.momentum
+    <Quantity 1.32521403e-34 J s / m>
+    >>> photon.frequency
+    <Quantity 59958491.6 1 / s>
+    >>> photon.spin
+    1
+    """
+
+    def __init__(self, quantity: Quantity = None):
+        self._set_default_photon_properties()
+        if (quantity is not None) and (not self._validate_quantity(quantity)):
+            raise ValueError(
+                "Specify non negative quantity with units of Energy, wavelength, frequency or momentum."
+            )
+
+        if quantity is None:
+            self.energy = None
+            self.frequency = None
+            self.momentum = None
+            self.wavelength = None
+        else:
+            if self._isMomentum(quantity):
+                self.momentum = quantity.to(u.kg * u.m / u.s)
+                self.wavelength = const.h / self.momentum
+                self.frequency = const.c / self.wavelength
+                self.energy = const.h * self.frequency
+            else:
+                self.energy = quantity.to(u.J, equivalencies=u.spectral())
+                self.frequency = quantity.to(u.s**-1, equivalencies=u.spectral())
+                self.wavelength = quantity.to(u.m, equivalencies=u.spectral())
+                self.momentum = const.h / self.wavelength
+
+    def _set_default_photon_properties(self):
+        """Set default properties of a Photon."""
+        self._mass = 0 * u.kg
+        self._charge = 0 * u.C
+        self._charge_number = 0
+        self._spin = 1
+        self._half_life = np.inf * u.s
+        self._allowed_units = dict(
+            {
+                "Energy": u.J,
+                "Wavelength": u.m,
+                "Frequency": u.s**-1,
+                "Momentum": u.kg * u.m / u.s,
+            }
+        )
+
+    def _validate_quantity(self, quantity: Quantity):
+        """
+        Checks if the type of ``quantity`` is one of the ``_allowed_units`` and
+        validates it's a valid non negative |Quantity|.
+        """
+        allowed_physical_types = [
+            u.get_physical_type(x) for x in self._allowed_units.values()
+        ]
+        isValidQuantity = quantity.unit.physical_type in allowed_physical_types
+        isNonNegativeQuantity = quantity.value > 0
+        return isValidQuantity and isNonNegativeQuantity
+
+    def _isMomentum(self, quantity: Quantity):
+        """Checks if the ``quantity`` is Momentum."""
+        return quantity.unit.physical_type == u.get_physical_type(
+            self._allowed_units["Momentum"]
+        )
+
```

Comment:
```suggestion

    def categories(self) -> Set[str]:
        """The categories of the |Photon|."""
        return self._categories

```
...to increase the similarity between `Photon` and `Particle`.